pkg/endpoint/endpoint_status.go

```diff
@@ -0,0 +1,228 @@
+// Copyright 2019 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package endpoint
+
+import (
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/cilium/cilium/api/v1/models"
+	"github.com/cilium/cilium/pkg/identity"
+	identitycache "github.com/cilium/cilium/pkg/identity/cache"
+	cilium_v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/policy/trafficdirection"
+)
+
+func getEndpointStatusControllers(status *models.EndpointStatus) (controllers cilium_v2.ControllerList) {
+	for _, c := range status.Controllers {
+		if c.Status == nil {
+			continue
+		}
+
+		if c.Status.ConsecutiveFailureCount > 0 {
+			s := cilium_v2.ControllerStatus{
+				Configuration: c.Configuration,
+				Name:          c.Name,
+				UUID:          string(c.UUID),
+				Status: cilium_v2.ControllerStatusStatus{
+					ConsecutiveFailureCount: c.Status.ConsecutiveFailureCount,
+					FailureCount:            c.Status.FailureCount,
+					LastFailureMsg:          c.Status.LastFailureMsg,
+					LastFailureTimestamp:    c.Status.LastFailureTimestamp.String(),
+					LastSuccessTimestamp:    c.Status.LastSuccessTimestamp.String(),
+					SuccessCount:            c.Status.SuccessCount,
+				},
+			}
+			if controllers == nil {
+				controllers = cilium_v2.ControllerList{s}
+			} else {
+				controllers = append(controllers, s)
+			}
+		}
+
+		if controllers != nil {
+			controllers.Sort()
+		}
+	}
+	return
+}
+
+func getEndpointStatusLog(e *Endpoint) (log []*models.EndpointStatusChange) {
```

Comment:
Any reason why it isn't an endpoint method?

Reply:
Good point, changed.